apps/condo/domains/user/integration/passport/routes.js

```diff
@@ -0,0 +1,135 @@
+            }
+            const Strategy = KNOWN_STRATEGIES[strategy]
+            const routes = this.generateRoutesByName(name)
+            this.#strategiesByNames[name] = new Strategy(strategyConfig, routes)
+        }
+    }
+
+    static onAuthFail (err, req, res, next) {
+        const gqlErr = wrapWithGQLError(err, { req }, ERRORS.AUTHORIZATION_FAILED)
+        next(gqlErr)
+    }
+
+    generateRoutesByName (name) {
+
+        return {
+            authURL: new URL(`${this.apiPrefix}/${name}`, conf['SERVER_URL']).toString(),
+            callbackURL: new URL(`${this.apiPrefix}/${name}/callback`, conf['SERVER_URL']).toString(),
+        }
+    }
+
+    addPassportRoutes (app, keystone) {
+        app.use(passport.initialize())
+
+        passport.serializeUser((user, done) => done(null, user.id))
+        passport.deserializeUser(async (id, done) => {
+            const context = await keystone.createContext({ skipAccessControl: true })
+            const user = await User.getOne(context, { id })
+            const err = user ? null : new Error('invalid user id to deserialize')
+            return done(err, user)
+        })
+
+        for (const [name, strategy] of Object.entries(this.#strategiesByNames)) {
+            passport.use(name, strategy.build())
+        }
+
+        const strategies = this.#strategiesByNames
+        function authenticate (req, res, next) {
+            const { provider } = req.params
+            if (!provider || !strategies.hasOwnProperty(provider)) {
+                return next(new GQLError({ ...ERRORS.UNKNOWN_PROVIDER, messageInterpolation: { provider } }, { req }))
+            }
+
+            passport.authenticate(provider)(req, res, next)
+        }
+
+        async function onAuthSuccess (req, res, next) {
+            const user = req.user
+
+            if (!user || !user.id) {
+                return next(new GQLError(ERRORS.AUTHORIZATION_FAILED, { req }))
+            }
+
+            try {
+                await keystone._sessionManager.startAuthedSession(req, {
+                    item: { id: user.id },
+                    list: keystone.lists['User'],
+                })
+
+                return res.redirect('/')
+            } catch (err) {
+                next(err)
+            }
+        }
+
+        app.get(
+            `${this.apiPrefix}/:provider`,
+            captureUserType,
+            authenticate,
+            PassportAuthRouter.onAuthFail,
+        )
+        app.get(
+            `${this.apiPrefix}/:provider/callback`,
+            ensureUserType,
+            authenticate,
+            onAuthSuccess,
+            PassportAuthRouter.onAuthFail,
+        )
+    }
+
+    getIdentityProviders () {
+        const providers = []
+
+        for (const strategy of Object.values(this.#strategiesByNames)) {
+            providers.push(...strategy.getProviders())
+        }
+
+        return providers
+    }
+
+    static init () {
```

Comment:
Static `init` and constructor in one class? Hard to understand so

Reply:
That's a pretty common pattern in other languages. you can create new instance with new keyword (usually private constructor) and initialise instance with ClassName.New (for instance), ClassName.Init (for global instance), or CreateNewClassName (via global functions), however in js there's no such term as a private constructor.

I can basically not init at all and everything will'be just fine. It's used just to optimise memory consumption.

So I guess its ok